experimental/sketching.py

```diff
@@ -0,0 +1,166 @@
+from jax import random
+from jax import numpy as np
+from neural_tangents._src.utils import dataclasses
+from typing import Optional, Callable
+
+
+# TensorSRHT of degree 2. This version allows different input vectors.
+@dataclasses.dataclass
+class TensorSRHT2:
+
+  input_dim1: int
+  input_dim2: int
+  sketch_dim: int
+
+  rng: np.ndarray
+  shape: Optional[np.ndarray] = None
+
+  rand_signs1: Optional[np.ndarray] = None
+  rand_signs2: Optional[np.ndarray] = None
+  rand_inds1: Optional[np.ndarray] = None
+  rand_inds2: Optional[np.ndarray] = None
+
+  replace = ...  # type: Callable[..., 'TensorSRHT2']
+
+  def init_sketches(self) -> 'TensorSRHT2':
+    rng1, rng2, rng3, rng4 = random.split(self.rng, 4)
+    rand_signs1 = random.choice(rng1, 2, shape=(self.input_dim1,)) * 2 - 1
+    rand_signs2 = random.choice(rng2, 2, shape=(self.input_dim2,)) * 2 - 1
+    rand_inds1 = random.choice(rng3,
+                               self.input_dim1,
+                               shape=(self.sketch_dim // 2,))
+    rand_inds2 = random.choice(rng4,
+                               self.input_dim2,
+                               shape=(self.sketch_dim // 2,))
+    shape = (self.input_dim1, self.input_dim2, self.sketch_dim)
+    return self.replace(shape=shape,
+                        rand_signs1=rand_signs1,
+                        rand_signs2=rand_signs2,
+                        rand_inds1=rand_inds1,
+                        rand_inds2=rand_inds2)
+
+  def sketch(self, x1, x2):
+    x1fft = np.fft.fftn(x1 * self.rand_signs1, axes=(-1,))[:, self.rand_inds1]
+    x2fft = np.fft.fftn(x2 * self.rand_signs2, axes=(-1,))[:, self.rand_inds2]
+    out = np.sqrt(1 / self.rand_inds1.shape[-1]) * (x1fft * x2fft)
+    return np.concatenate((out.real, out.imag), 1)
+
+# Standard SRHT as a function
+def standardsrht(x, rand_inds, rand_signs):
+    xfft = np.fft.fftn(x * rand_signs, axes=(-1,))[:, rand_inds]
+    return np.sqrt(1 / rand_inds.shape[0]) * xfft
+
+@dataclasses.dataclass
+class PolyTensorSketch:
+    
+    rng: np.ndarray
+    
+    input_dim: int
+    sketch_dim: int
+    degree: int
+    
+    tree_rand_signs: Optional[list] = None
+    tree_rand_inds: Optional[list] = None
+    rand_signs: Optional[np.ndarray] = None
+    rand_inds: Optional[np.ndarray] = None
+    
+    replace = ...
```

Comment:
Is this ever used? If it is, let's annotate it as `replace = ...  # type: Callable[..., 'PolyTensorSketch']` for pytype checks to work (if not, let's remove it, as well as other `replace` attributes in other classes?).